src/visit.rs

```diff
@@ -0,0 +1,211 @@
+use std::collections::BTreeMap;
+
+use crate::{FeatureValue, Metadata, Package};
+
+/// A visitor over a package's features and their dependencies.
+pub trait FeatureVisitor {
+    /// The error type of a walk.
+    type Error;
+
+    /// Visits a missing dependency.
+    ///
+    /// Return `Ok(())` to continue the walk, or `Err(…)` to abort it.
+    fn visit_missing_dependency(&mut self, dep_name: &str) -> Result<(), Self::Error>;
+
+    /// Visits a missing package.
+    ///
+    /// Return `Ok(())` to continue the walk, or `Err(…)` to abort it.
+    fn visit_missing_package(&mut self, pkg_name: &str) -> Result<(), Self::Error>;
+
+    /// Visits a feature that's enabling another feature.
+    ///
+    /// Return `Ok(<walk>)` where `<walk>` indicates whether or not to walk the feature's downstream dependencies,
+    /// or `Err(…)` to abort the walk.
+    fn visit_feature(
+        &mut self,
+        package: &Package,
+        feature_name: &str,
+    ) -> Result<bool, Self::Error> {
+        let (..) = (package, feature_name);
+        Ok(true)
+    }
+
+    /// Visits a feature that's enabling a dependency with `dep:dep_name` syntax.
+    ///
+    /// Return `Ok(<walk>)` where `<walk>` indicates whether or not to walk the feature's downstream dependencies,
+    /// or `Err(…)` to abort the walk.
+    fn visit_dep(&mut self, package: &Package, dep_name: &str) -> Result<bool, Self::Error> {
+        let (..) = (package, dep_name);
+        Ok(true)
+    }
+
+    /// Visits a feature that's enabling a feature on a dependency with `crate_name/feat_name` syntax.
+    ///
+    /// Return `Ok(<walk>)` where `<walk>` indicates whether or not to walk the feature's downstream dependencies,
+    /// or `Err(…)` to abort the walk.
+    fn visit_dep_feature(
+        &mut self,
+        package: &Package,
+        dep_name: &str,
+        dep_feature: &str,
+        weak: bool,
+    ) -> Result<bool, Self::Error> {
```

Comment:
The example syntax is `crate_name/feat_name` but neither of those names appears in the list of function arguments. Could you make the example match the function arguments?

Reply:
Fixed.